Mathlib/Algebra/Lie/Weights/Killing.lean

```diff
@@ -351,6 +396,86 @@ lemma isCompl_ker_weight_span_coroot (α : Weight K H L) :
     replace hα : corootSpace α ≠ ⊥ := by simpa using hα
     rwa [ne_eq, ← LieSubmodule.coe_toSubmodule_eq_iff] at hα
 
+lemma exists_isSl2Triple_of_weight_isNonZero {α : Weight K H L} (hα : α.IsNonZero) :
+    ∃ h e f : L, IsSl2Triple h e f ∧ e ∈ rootSpace H α ∧ f ∈ rootSpace H (- α) := by
+  obtain ⟨e, heα : e ∈ rootSpace H α, he₀ : e ≠ 0⟩ := α.exists_ne_zero
+  obtain ⟨f, hfα, hf⟩ : ∃ f ∈ rootSpace H (-α), killingForm K L e f ≠ 0 := by
+    contrapose! he₀
+    simpa using mem_ker_killingForm_of_mem_rootSpace_of_forall_rootSpace_neg K L H heα he₀
+  have hef := lie_eq_killingForm_smul_of_mem_rootSpace_of_mem_rootSpace_neg heα hfα
+  let h : H := ⟨⁅e, f⁆, hef ▸ Submodule.smul_mem _ _ (Submodule.coe_mem _)⟩
+  have hh : α h ≠ 0 := by
+    change (α : H →ₗ[K] K) h ≠ 0
+    have : h = killingForm K L e f • (cartanEquivDual H).symm α := by simp [Subtype.ext_iff, hef]
+    rw [this, map_smul, smul_eq_mul, ne_eq, mul_eq_zero, not_or]
+    exact ⟨hf, root_apply_cartanEquivDual_symm_ne_zero hα⟩
+  let f' := (2 * (α h)⁻¹) • f
+  replace hef : ⁅⁅e, f'⁆, e⁆ = 2 • e := by
+    have : ⁅⁅e, f⁆, e⁆ = α h • e := lie_eq_smul_of_mem_rootSpace heα h
+    rw [lie_smul, smul_lie, this, ← smul_assoc, smul_eq_mul, mul_assoc, inv_mul_cancel hh,
+      mul_one, two_smul, two_smul]
+  refine ⟨⁅e, f'⁆, e, f', ⟨fun contra ↦ ?_, rfl, hef, ?_⟩, heα, Submodule.smul_mem _ _ hfα⟩
+  · rw [contra] at hef
+    have _i : NoZeroSMulDivisors ℤ L := NoZeroSMulDivisors.int_of_charZero K L
+    simp only [zero_lie, eq_comm (a := (0 : L)), smul_eq_zero, OfNat.ofNat_ne_zero, false_or] at hef
+    contradiction
+  · have : ⁅⁅e, f⁆, f⁆ = - α h • f := lie_eq_smul_of_mem_rootSpace hfα h
+    rw [lie_smul, lie_smul, smul_lie, this, ← smul_assoc, smul_eq_mul, mul_assoc]
+    simp only [neg_smul, smul_neg, ← smul_assoc, smul_eq_mul, mul_assoc, neg_inj, f']
+    field_simp
+
+lemma isSl2Triple_h_eq_coroot {α : Weight K H L} (hα : α.IsNonZero)
+    {h e f : L} (ht : IsSl2Triple h e f) (heα : e ∈ rootSpace H α) (hfα : f ∈ rootSpace H (- α)) :
+    h = coroot α := by
+  have hef := lie_eq_killingForm_smul_of_mem_rootSpace_of_mem_rootSpace_neg heα hfα
+  let h' : H := ⟨h, by rw [← ht.lie_e_f, hef]; exact Submodule.smul_mem _ _ (Submodule.coe_mem _)⟩
+  suffices h' = coroot α by rw [← this]
```

Comment:
How about
```suggestion
  lift h to H using by simpa only [← ht.lie_e_f, hef] using H.smul_mem _ (Submodule.coe_mem _)
  congr 1
```
and then using `h` everywhere instead of `h'`?